### PR TITLE
feat: add maxAge parameter to limit the staleness of a historical data point

### PR DESCRIPTION
**Pull request by mrice32**
One Oval-related trust vector has surfaced recently. If the oracle is not responsible for keeping snapshots, then someone else is. If that entity allows snapshots to become sufficiently stale, it can lead to the application using very stale prices when a new price comes in.

One idea was to add a maxAge that limits how stale a historical snapshot can be before the current data is used instead.

Note: this PR is an example of how this would work. Tests need to be fixed and added.

### CODE REVIEW

File: src/controllers/BaseController.sol [hunk 1]
```diff
@@ -83,4 +94,11 @@ abstract contract BaseController is Ownable, Oval {
     function maxTraversal() public view override returns (uint256) {
         return maxTraversal_;
     }
+
+    /**
+     * @notice Max age of a historical price that can be used instead of the current price.
+     */
+    function maxAge() public view override returns (uint256) {
+        return maxTraversal_;
```
Comment (chrismaree) [inline src/controllers/BaseController.sol:102]: a critical error here! should be `maxAge_`. found while fixing tests for CI

File: test/fork/adapters/ChainlinkSourceAdapter.sol [hunk 2]
```diff
@@ -8,18 +8,8 @@ import {ChainlinkSourceAdapter} from "../../../src/adapters/source-adapters/Chai
 import {DecimalLib} from "../../../src/adapters/lib/DecimalLib.sol";
 import {IAggregatorV3Source} from "../../../src/interfaces/chainlink/IAggregatorV3Source.sol";
 
-contract TestedSourceAdapter is ChainlinkSourceAdapter {
+contract TestedSourceAdapter is ChainlinkSourceAdapter, BaseController {
```
Comment (chrismaree) [inline test/fork/adapters/ChainlinkSourceAdapter.sol:11]: I had to make a somewhat opinionated change to this test (as well as all the other fork tests) to get them to work with the `maxAge` under situations where we do multi-round traversal:

due to the previous inheritance structure, we had `maxAge` was set to 0. we have a number of tests that look back over a number of rounds. in conjunction with this 0 value this broke some of the tests. our options to fix this were:
a) implement the `TestedSourceAdapter` to return a value for `maxAge` to something meaningful (like `1 days`) or 
b) add `BaseController` to the inheritance path for the `TestedSourceAdapter` to get the functionality of (a) for free. the upside with this is we also ditch all the `public view override` in all the unit tests.

I opted for (b).

File: test/fork/adapters/ChainlinkSourceAdapter.sol [hunk 3]
```diff
@@ -107,6 +97,33 @@ contract ChainlinkSourceAdapterTest is CommonTest {
         assertTrue(uint256(roundId) == lookBackRoundId);
     }
 
+    function testCorrectlyBoundsMaxLooBackByMaxAge() public {
```
Comment (chrismaree) [inline test/fork/adapters/ChainlinkSourceAdapter.sol:100]: added some tests to show maxAge is respected. some done in the snapshotSource & permissions tests added to controller tests.

**Review by Reinis-FRP — APPROVED** [review]

File: src/adapters/source-adapters/ChainlinkSourceAdapter.sol [hunk 4]
```diff
@@ -94,7 +94,9 @@ abstract contract ChainlinkSourceAdapter is DiamondRootOval {
             _searchRoundDataAt(timestamp, roundId, maxTraversal);
 
         // Validate returned data. If it is uninitialized we fallback to returning the current latest round data.
-        if (historicalUpdatedAt > 0) return (historicalAnswer, historicalUpdatedAt, historicalRoundId);
+        if (historicalUpdatedAt > block.timestamp - maxAge()) {
```
Comment (md0x) [inline src/adapters/source-adapters/ChainlinkSourceAdapter.sol:97]: Question: Can we leave the `historicalUpdatedAt > 0` in this particular source adapter for gas savings given that CL has history?

**Review by md0x — APPROVED** [review]
Looks good! Added a question related to a potential gas saving in CL source adapter.

**Review by chrismaree — APPROVED** [review]

File: src/controllers/BaseController.sol (hunk 1 above)
Comment (mrice32) [reply src/controllers/BaseController.sol]: Amazing!

File: test/fork/adapters/ChainlinkSourceAdapter.sol (hunk 2 above)
Comment (mrice32) [reply test/fork/adapters/ChainlinkSourceAdapter.sol:11]: Perfect, I agree that's a lot better and much less complicated/copypasta.

File: src/adapters/source-adapters/ChainlinkSourceAdapter.sol (hunk 4 above)
Comment (mrice32) [reply src/adapters/source-adapters/ChainlinkSourceAdapter.sol:97]: I think it's ideal for maxAge to be respected regardless of the source. It's a lot easier to reason about the system as a whole in that case. It should add only minor additional gas, as maxAge will typically be an immutable variable.

**Pull request merged**